### PR TITLE
feat: support runtime value and state management across sub-workflows

### LINKED ISSUE #51: Support state management across workflow, sub-workflow and steps
We need to pass the state from workflow to sub-workflow/steps. A copy of the parent workflow needs to be passed so that parent workflow isn't modified by the children. A parent however can mutate its state and pass different versions of its state to different sub-workflow/step as required.

We need to ensure rollback operation also mutates the state (unwind the changes as required)

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description
This pull request introduces significant improvements to workflow state management, particularly around deep cloning of state for sub-workflows and robust error handling during cloning. It also adds a `.testcoverage.yml` configuration for coverage enforcement, and updates interfaces and tests to support the new cloning semantics. The most important changes are grouped below:

**Workflow State Management and Cloning Improvements:**

* The `StateBag` interface's `Clone()` method now returns `(StateBag, error)` to support error reporting during deep cloning. The `SyncStateBag` implementation attempts a deep clone for items that implement a `Clone` method (with or without error), falling back to a shallow copy otherwise. Errors during cloning are correctly propagated. (`automa.go`, `state.go`, [[1]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83L25-R26) [[2]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aL27-R75)
* Workflow execution now clones the workflow state for sub-workflows, ensuring that sub-workflows cannot mutate the parent workflow's state unintentionally. If state cloning fails, the step is marked as failed and not executed. Rollbacks use per-step state snapshots to ensure deterministic behavior. (`workflow.go`, [[1]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR10-R37) [[2]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR54-R63) [[3]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL67-R138) [[4]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL124-R208) [[5]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR222-R273) [[6]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR287-R298)

**Interface and API Enhancements:**

* The `Step` interface gains a `WithState(StateBag) Step` method, and both `defaultStep` and `workflow` implement this to support explicit state assignment. (`automa.go`, `step_default.go`, `workflow.go`, [[1]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83R14) [[2]](diffhunk://#diff-51ba6897a18f3ca97dd7dc56fcc91ddc51081ad6aff5e62e43060edb49542eb2R30-R39) [[3]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR54-R63)
* The `StepBuilder.BuildAndCopy()` method now handles errors from state cloning and returns an error if cloning fails. (`step_builder.go`, [step_builder.goR112-R117](diffhunk://#diff-de7fbc5fe7dcf3fe1020642b6d29bc5047a74e96a7db21e1157221bc840cc214R112-R117))

**Testing and Coverage:**

* Extensive new tests verify deep and shallow cloning behavior, error propagation from failed clones, and handling of nil receivers. (`state_test.go`, [[1]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cL159-L174) [[2]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cR231-R343)
* Added `.testcoverage.yml` for Go test coverage configuration, with thresholds and exclusion rules. (`.testcoverage.yml`, [.testcoverage.ymlR1-R53](diffhunk://#diff-6d09ae139b77769862a6695df9e82e6f27d74fa6ace5b4ee56bfc51272458b7aR1-R53))

**Dependency and Import Updates:**

* Added `golang.org/x/sync` dependency and improved import organization. (`go.mod`, [[1]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R9) [[2]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aR5-R8) [[3]](diffhunk://#diff-de7fbc5fe7dcf3fe1020642b6d29bc5047a74e96a7db21e1157221bc840cc214R6) [[4]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cR7-R9)

(References: [[1]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83R14) [[2]](diffhunk://#diff-2f78a604a2a388d0cd14602cb33e14d37d22b5a98b276f6774a39cbcd2c49d83L25-R26) [[3]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aL27-R75) [[4]](diffhunk://#diff-51ba6897a18f3ca97dd7dc56fcc91ddc51081ad6aff5e62e43060edb49542eb2R30-R39) [[5]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR10-R37) [[6]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR54-R63) [[7]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL67-R138) [[8]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL124-R208) [[9]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR222-R273) [[10]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR287-R298) [[11]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL197-R312) [[12]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL218-L221) [[13]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL271-R384) [[14]](diffhunk://#diff-de7fbc5fe7dcf3fe1020642b6d29bc5047a74e96a7db21e1157221bc840cc214R112-R117) [[15]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cL159-L174) [[16]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cR231-R343) [[17]](diffhunk://#diff-6d09ae139b77769862a6695df9e82e6f27d74fa6ace5b4ee56bfc51272458b7aR1-R53) [[18]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R9) [[19]](diffhunk://#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aR5-R8) [[20]](diffhunk://#diff-de7fbc5fe7dcf3fe1020642b6d29bc5047a74e96a7db21e1157221bc840cc214R6) [[21]](diffhunk://#diff-864944c880a2e4db1b8f93f33a58303d840fb4b6c50147cc2f915c6d48f7bf2cR7-R9)

### Related Issues

* Closes #51
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: workflow.go [hunk 1]
```diff
@@ -137,24 +219,58 @@ func (w *workflow) Execute(ctx context.Context) *Report {
...
+		}
+
+		// attach snapshot for possible rollback (keeps alignment even if nil)
+		stepStates = append(stepStates, stepState)
+
+		// make sure step has its state before calling Prepare so Prepare can access it
+		step = step.WithState(stepState)
+
+		// prepare step context
+		if statePrepError == nil {
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go]: It will never go to line 298 if there is any error during state or ctx preparation. However, we are now initializing stepCtx = ctx
